core/src/test/scala/unit/kafka/server/ClientQuotaManagerTest.scala

```diff
@@ -38,43 +39,171 @@ class ClientQuotaManagerTest {
-      assertEquals("Should return the default value (500)", new Quota(500, true), clientMetrics.quota("p1"))
+      clientMetrics.updateQuota(client1.configUser, client1.configClientId, Some(new Quota(500, true)))
+      assertEquals("Should return the default value (500)", new Quota(500, true), clientMetrics.quota(client1.user, client1.clientId))
 
-      throttleTimeMs = clientMetrics.recordAndMaybeThrottle("p1", 0, this.callback)
+      throttleTimeMs = clientMetrics.recordAndMaybeThrottle(client1.user, client1.clientId, 0, this.callback)
       assertTrue(s"throttleTimeMs should be > 0. was $throttleTimeMs", throttleTimeMs > 0)
+
+      // Case 4: Set high default quota, remove p1 quota. p1 should no longer be throttled
+      clientMetrics.updateQuota(client1.configUser, client1.configClientId, None)
+      clientMetrics.updateQuota(defaultConfigClient.configUser, defaultConfigClient.configClientId, Some(new Quota(4000, true)))
+      assertEquals("Should return the newly overridden value (4000)", new Quota(4000, true), clientMetrics.quota(client1.user, client1.clientId))
+
+      throttleTimeMs = clientMetrics.recordAndMaybeThrottle(client1.user, client1.clientId, 0, this.callback)
+      assertEquals(s"throttleTimeMs should be 0. was $throttleTimeMs", 0, throttleTimeMs)
+
     } finally {
       clientMetrics.shutdown()
     }
   }
 
   @Test
+  def testClientIdQuotaParsing() {
+    val client1 = UserClient("ANONYMOUS", "p1", None, Some("p1"))
+    val client2 = UserClient("ANONYMOUS", "p2", None, Some("p2"))
+    val randomClient = UserClient("ANONYMOUS", "random-client-id", None, None)
+    val defaultConfigClient = UserClient("", "", None, Some(ConfigEntityName.Default))
+    testQuotaParsing(config, client1, client2, randomClient, defaultConfigClient)
+  }
+
+  @Test
+  def testUserQuotaParsing() {
+    val client1 = UserClient("User1", "p1", Some("User1"), None)
+    val client2 = UserClient("User2", "p2", Some("User2"), None)
+    val randomClient = UserClient("RandomUser", "random-client-id", None, None)
+    val defaultConfigClient = UserClient("", "", Some(ConfigEntityName.Default), None)
+    val config = ClientQuotaManagerConfig(quotaBytesPerSecondDefault = Long.MaxValue)
+    testQuotaParsing(config, client1, client2, randomClient, defaultConfigClient)
+  }
+
+  @Test
+  def testUserClientIdQuotaParsing() {
+    val client1 = UserClient("User1", "p1", Some("User1"), Some("p1"))
+    val client2 = UserClient("User2", "p2", Some("User2"), Some("p2"))
+    val randomClient = UserClient("RandomUser", "random-client-id", None, None)
+    val defaultConfigClient = UserClient("", "", Some(ConfigEntityName.Default), Some(ConfigEntityName.Default))
+    val config = ClientQuotaManagerConfig(quotaBytesPerSecondDefault = Long.MaxValue)
+    testQuotaParsing(config, client1, client2, randomClient, defaultConfigClient)
+  }
+
+  @Test
+  def testUserQuotaParsingWithDefaultClientIdQuota() {
```

Comment:
Could you add a comment to explain how this is different from testUserQuotaParsing()?